test/extended/networking/cnimigration.go

```diff
@@ -0,0 +1,326 @@
+		return fmt.Errorf("timed out waiting for network operator to acknowledge migration is in-progress: %v", err)
+	}
+	migrationStart := time.Now()
+	if err := wait.PollUntilContextTimeout(ctx, 10*time.Second, maxMigrationDuration, true, func(ctx context.Context) (bool, error) {
+		isMigrationCompleted, err := isMigrationComplete(ctx, c)
+		if err != nil {
+			return false, err
+		}
+		return isMigrationCompleted, nil
+	}); err != nil {
+		return fmt.Errorf("cluster did not complete CNI migration from CNI %s to %s: %v", initCNI, targetCNI, err)
+	}
+	migrationEnd := time.Now()
+	migrationDuration := migrationEnd.Sub(migrationStart)
+	migrationDurationMinutes := migrationDuration.Minutes()
+	framework.Logf("Completed migration from CNI %s to %s. Completed in %0.2f minutes", initCNI, targetCNI, migrationDurationMinutes)
+	recordEvent(kubeClient, uid, "Migration", "MigrationSucceeded",
+		fmt.Sprintf("from CNI %s to %s. Completed in %0.2f minutes", initCNI, targetCNI, migrationDurationMinutes), false)
+	return nil
+}
+
+// recordEvent attempts to record an event to the cluster to indicate actions taken during a live
+// migration for timeline review.
+func recordEvent(client kubernetes.Interface, uid, action, reason, note string, warning bool) {
+	currentTime := metav1.MicroTime{Time: time.Now()}
+	t := corev1.EventTypeNormal
+	if warning {
+		t = corev1.EventTypeWarning
+	}
+	ctx, cancelFn := context.WithTimeout(context.TODO(), 10*time.Second)
+	defer cancelFn()
+	ns := "openshift-network-operator"
+	_, err := client.EventsV1().Events(ns).Create(ctx, &eventsv1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%v.%x", "cni-migration", currentTime.UnixNano()),
+		},
+		Regarding:           corev1.ObjectReference{Kind: "Network", Name: networkCRName, Namespace: ns, APIVersion: configv1.GroupVersion.String()},
+		Action:              action,
+		Reason:              reason,
+		Note:                note,
+		Type:                t,
+		EventTime:           currentTime,
+		ReportingController: "openshift-tests.openshift.io/cni-migration",
+		ReportingInstance:   uid,
+	}, metav1.CreateOptions{})
+	if err != nil {
+		framework.Logf("Unable to record cluster event: %v", err)
+	}
+}
+
+func isSupportedCNI(cni string) bool {
+	return cni == OVNKubernetesPluginName || cni == OpenshiftSDNPluginName
+}
+
+func getRollBackCNI(cni string) string {
+	if cni == OVNKubernetesPluginName {
+		return OpenshiftSDNPluginName
+	}
+	if cni == OpenshiftSDNPluginName {
+		return OVNKubernetesPluginName
+	}
+	panic(fmt.Sprintf("unsupported CNI %q specified. Unable to determine rollback CNI", cni))
+}
+
+var _ = g.Describe("[sig-network][Feature:CNIMigration]", g.Ordered, func() {
+	var tc testConfig
+
+	g.BeforeAll(func() {
+		tc = getTestConfig()
+	})
+
+	g.BeforeEach(func() {
+		if tc.targetCNI == "" {
+			e2eskipper.Skipf("CNI migration tests are disabled because %s environment key does not have a value which contains 'target-key=$CNI'", optionsEnvKey)
```

Comment:
Just a quick check, do you want a return here? Looks like isSupportedCNI could still be run and return false, leading to a failed test when you wanted a skip.  Actually do you want to perhaps fail if we get to this point?

Reply:
Aha all good, thanks.